go-controller/pkg/cni/helper_linux_test.go

```diff
@@ -0,0 +1,987 @@
+			} else {
+				assert.Nil(t, err)
+			}
+			mockNetLinkOps.AssertExpectations(t)
+		})
+	}
+}
+
+func TestMoveIfToNetns(t *testing.T) {
+	mockNetLinkOps := new(util_mocks.NetLinkOps)
+	mockNetNS := new(cni_ns_mocks.NetNS)
+	mockLink := new(netlink_mocks.Link)
+	// below sets the `netLinkOps` in util/net_linux.go to a mock instance for purpose of unit tests execution
+	util.SetNetLinkOpMockInst(mockNetLinkOps)
+
+	tests := []struct {
+		desc                 string
+		inpIfaceName         string
+		inpNetNs             ns.NetNS
+		errMatch             error
+		netLinkOpsMockHelper []ovntest.TestifyMockHelper
+		netNsOpsMockHelper   []ovntest.TestifyMockHelper
+	}{
+		{
+			desc:         "test code path when LinkByName() returns error",
+			inpIfaceName: "testIfaceName",
+			inpNetNs:     nil,
+			errMatch:     fmt.Errorf("failed to lookup vf device"),
+			netLinkOpsMockHelper: []ovntest.TestifyMockHelper{
+				{"LinkByName", []string{"string"}, []interface{}{nil, fmt.Errorf("mock error")}},
+			},
+		},
+		{
+			desc:         "test code path when LinkSetNsFd() returns error",
+			inpIfaceName: "testIfaceName",
+			inpNetNs:     mockNetNS,
+			errMatch:     fmt.Errorf("failed to move device"),
+			netLinkOpsMockHelper: []ovntest.TestifyMockHelper{
+				{"LinkByName", []string{"string"}, []interface{}{mockLink, nil}},
+				{"LinkSetNsFd", []string{"*mocks.Link", "int"}, []interface{}{fmt.Errorf("mock error")}},
+			},
+			netNsOpsMockHelper: []ovntest.TestifyMockHelper{
+				{"Fd", []string{}, []interface{}{uintptr(123456)}},
+			},
+		},
+		{
+			desc:         "test success path",
+			inpIfaceName: "testIfaceName",
+			inpNetNs:     mockNetNS,
+			netLinkOpsMockHelper: []ovntest.TestifyMockHelper{
+				{"LinkByName", []string{"string"}, []interface{}{mockLink, nil}},
+				{"LinkSetNsFd", []string{"*mocks.Link", "int"}, []interface{}{nil}},
+			},
+			netNsOpsMockHelper: []ovntest.TestifyMockHelper{
+				{"Fd", []string{}, []interface{}{uintptr(123456)}},
+			},
+		},
+	}
+	for i, tc := range tests {
+		t.Run(fmt.Sprintf("%d:%s", i, tc.desc), func(t *testing.T) {
+			for _, item := range tc.netLinkOpsMockHelper {
+				call := mockNetLinkOps.On(item.OnCallMethodName)
+				for _, arg := range item.OnCallMethodArgType {
+					call.Arguments = append(call.Arguments, mock.AnythingOfType(arg))
+				}
+				for _, ret := range item.RetArgList {
+					call.ReturnArguments = append(call.ReturnArguments, ret)
+				}
+				call.Once()
+			}
+			for _, item := range tc.netNsOpsMockHelper {
+				call := mockNetNS.On(item.OnCallMethodName)
+				for _, arg := range item.OnCallMethodArgType {
+					call.Arguments = append(call.Arguments, mock.AnythingOfType(arg))
+				}
+				for _, ret := range item.RetArgList {
+					call.ReturnArguments = append(call.ReturnArguments, ret)
+				}
+				call.Once()
+			}
+			err := moveIfToNetns(tc.inpIfaceName, tc.inpNetNs)
+			t.Log(err)
+			if tc.errMatch != nil {
+				assert.Contains(t, err.Error(), tc.errMatch.Error())
```

Comment:
Can we just it errors out or not? The error string is not a contract between the API caller and the API implementor. Also, you are only checking some part of the error message. So, it is better to just check if the API failed or not.

Reply:
My initial approach was like what you were suggesting, however, the current approach taken is after I got a comment (https://github.com/ovn-org/ovn-kubernetes/pull/1474#discussion_r447723068) from @danwinship to one of my PRs related to tests for annotations files in pkg/util. If we just checked for error only, we could still accidentally make tests pass and not catch any changes made to the code.